File: ui/src/core/model/lobby.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { getLobby } from '@/commands/lobby';

export class LobbyImpl implements Lobby {
  public readonly ready: readonly PlayerId[];

  private constructor(lobby: Lobby) {
    this.ready = lobby.ready;
  }

  public isReady(id: PlayerId) {
    return this.ready.includes(id);
  }

  public static create(lobby: Lobby) {
    return new LobbyImpl(lobby);
  }

  public static async load() {
    const lobby = await getLobby();
    return this.create(lobby);
  }
}

File: crates/nil/src/command/lobby.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::error::Result;
use crate::manager::ManagerExt;
use nil_core::lobby::LobbyState;
use nil_core::player::PlayerId;
use tauri::AppHandle;

#[tauri::command]
pub async fn get_lobby(app: AppHandle) -> Result<LobbyState> {
  app
    .client(async |cl| cl.get_lobby().await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn set_lobby_ready(app: AppHandle, id: PlayerId, ready: bool) -> Result<()> {
  app
    .client(async |cl| cl.set_lobby_ready(id, ready).await)
    .await?
    .map_err(Into::into)
}
